twisted/python/zippath.py

```diff
@@ -128,9 +129,11 @@ def listdir(self):
             if self.isdir():
                 return list(self.archive.childmap[self.pathInArchive].keys())
             else:
-                raise OSError(errno.ENOTDIR, "Leaf zip entry listed")
+                raise UnlistableError(
+                    OSError(errno.ENOTDIR, "Leaf zip entry listed"))
         else:
-            raise OSError(errno.ENOENT, "Non-existent zip entry listed")
+            raise UnlistableError(
```

Comment:
Some explanation:

[twisted.python.filepath.UnlistableError](https://github.com/twisted/twisted/blob/trunk/twisted/python/filepath.py#L230-L248) takes an `OSError` instance as its first argument.  While this looks strange, it's the right thing to do.

Also, it appears that `OSError(errno.ENOENT, ...)` and `OSError(errno.ENOTDIR, ...)` are raised as non-`OSError` exceptions on Windows, which is interesting.